Avoid multiclicking in Admin and General settings

grzesiek2010 opened the pull request:
Closes #3713 

I built this pr on top of #3699 to test the newly factored `MultiClickGuard` class.

#### What has been done to verify that this works as intended?
I tested the implemented changes manually.

#### Why is this the best possible solution? Were any other approaches considered?
We just need to use our MultiClickGuard class to avoid such bugs since android doesn't do that for us.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
I didn't only fix the case described in the issue with the Admin Password dialog. I decided to add the check for all options in Admin Settings and even in General Settings. I was able for example to reproduce the same issue with General Settings option that is in Admin Settings so I guess the same is possible with other options as well so I decided to handle all of them just in case. Please pay attention to that option (General Settings from Admin Settings) since in order to handle it as well I needed to add some more refactoring.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [ ] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [ ] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [ ] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

grzesiek2010 committed:
Factored out MultiClickGuard class

grzesiek2010 committed:
Avoid multiclicking in Admin preferences

grzesiek2010 committed:
Avoid multiclicking in General preferences

grzesiek2010 committed:
Handle starting general prefs from admin prefs in AdminPreferencesFragment to avoid multiclicking

seadowg commented:
Two things I'm thinking about: for this whole set of changes:

1. I'm wondering if we could have the helper let us setup click listeners rather than us having to always check in the click listener. In my head it would be nice if we could do something like this:

    ```java
    MultiClickGuard.setOnClickListener(button, view -> {
        // handle click            
    });
    ```

2. I'm a little worried this could create (or might have already created) problems for tests where multiple clicks might happen very quickly. I think we can ease that worry easily by disabling it somehow in both Espresso and Robolectric.

seadowg reviewed (commented):
As above!

seadowg reviewed (changes requested):
Just sticking this into "Request changes" as leaving a "Comment" review apparently doesn't remove the "Review required" state annoyingly. Feel free to re-request review if you disagree with my comments!

grzesiek2010 commented:
>'m wondering if we could have the helper let us setup click listeners rather than us having to always check in the click listener. In my head it would be nice if we could do something like this:
MultiClickGuard.setOnClickListener(button, view -> {
    // handle click            
});

I think it's not possible because we use MultiClickGuard in too many different contexts I mean:
- onClick();
- onPreferenceClick();
- clickImage();
- onItemClick();

In many cases we don't even set the listener but just implements and overrides like [here](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/activities/FormChooserListActivity.java#L123) for example. So it would require a lot of changes and the code won't be perfect either way.

>I'm a little worried this could create (or might have already created) problems for tests where multiple clicks might happen very quickly. I think we can ease that worry easily by disabling it somehow in both Espresso and Robolectric.

I think it is disabled https://github.com/opendatakit/collect/pull/3719/files#diff-b4b1a1c24bed33b17e60ef5611e6b38dR19

seadowg commented on collect_app/src/main/java/org/odk/collect/android/utilities/MultiClickGuard.java:
```diff
@@ -0,0 +1,27 @@
+        long elapsedRealtime = SystemClock.elapsedRealtime();
+        boolean isSameClass = className.equals(lastClickName);
+        boolean isBeyondThreshold = elapsedRealtime - lastClickTime > CLICK_DEBOUNCE_MS;
+        boolean isBeyondTestThreshold = lastClickTime == 0 || lastClickTime == elapsedRealtime; // just for tests
```
I'm not sure I understand this. Does this assume no time passes between clicks in tests? As far as I can see this is `true` if not time has passed since the last click. In both espresso and normal junit it's very possible for a millisecond to pass between clicks. Maybe I just need some more caffeine ☕️.

grzesiek2010 replied:
Yes, someone has added it on purpose so I guess it fixed the problem he had run into.

grzesiek2010 replied:
Generally it shouldn't be a problem in which test a situation like that could happen that you need to click on the same element twice quickly?

seadowg replied:
Is doesn't need to be the same element though right? It just needs to be in the same class (generally we're passing the Activity). Personlly the solution confuses me and I'm sure there are simpler ways to work out if we're in tests or not. I can look at later seeing as I have a personal vendetta against the line 😁. Let's just make sure to run `connectedAndroidTest` to check Espresso doesn't run into any problems with all the new guards.

grzesiek2010 replied:
> Is doesn't need to be the same element though right?

right.

seadowg commented:
> I think it's not possible because we use MultiClickGuard in too many different contexts

That's a great point! I agree we just go with the way it is for now.

kkrawczyk123 commented:
@grzesiek2010 I was able to open twice Form Metadata in User and device identity but only just once on Android 7.1. @mmarciniak90, please keep it in mind while testing as you have currently more devices than I do. 

Verified on Androids 5.1, 7.1 and 10.

Verified cases:
- multiclick on all settings in admin settings 
- multiclick on all settings in general settings 
- light mode/dark mode

grzesiek2010 commented:
Yeah actually I didn't dig that deep and handled just options from the main screens not their children... I'll take a look to check if it make sense.

mmarciniak90 commented:
> I was able to open twice Form Metadata in User and device identity but only just once on Android 7.1

I had more luck because I was able to reproduce it on Android 6.0, 7.0, 8.1, 9.0

grzesiek2010 committed:
Avoid multiclicking on Form Metadata

grzesiek2010 commented:
> I was able to open twice Form Metadata in User and device identity

ok I fixed that case too but I think it doesn't make sense to handle every single option (it's especially if there is no one place where I could do that), so that's the last case I fixed if you find something else what doesn't cause any problems apart from the fact it opens something twice please ignore.

mmarciniak90 commented:
**Tested with success**

Verified on Android 6.0, 7.0, 8.1, 9.0

Verified cases:
* multi-click on all settings in admin settings
* multi-click on all settings in general settings
* multi-click on Form Metadata in User and device identity
* light mode/dark mode

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"
